### ISSUE TITLE
V3 Punchlist

### ISSUE THREAD

**Issue by jxson** [jxson/front-matter#37]
Here is a meta list of items for v3, in no particular order. This is mostly a brain dump and some notes from chatting with @axdg 

* [ ] Refactor the tests so they work in browser and add a SauceLabs badge.
* [ ] Add benchmarks and add results to the README
* [ ] Refactor internal implementation to resemble axdg/dsfm (Faster Regex usage, etc.)
* [ ] Steaming support
* [ ] Custom parser support (maybe the only built it one is JSON, everything else should use this API to keep down dependency sizes?)
* [ ] Deprecate v2
* [ ] Clean up README 
  * [ ] Add information about the size of the module (including dependencies) for browser users
  * [ ] Add better information for collaborators and maintainers about development workflow (testing, code reviews, publishing, etc.)

**Comment by axdg** [comment]
Yep, this all looks good.

Just a few notes:

 - Regarding #26; @qimingweng I'm not sure that 'stringifying' a 'contents' object is necessarily something that needs to be done by this module. It's something that is trivial to do (just concatenating stringified attributes and the body of the document). Creating a stream that does this is also pretty trivial. While it's not a bad feature to have in the module, it doesn't really fit in with the idea of custom parsers, where the user supplies their own `load` function for the front-matter. It would mean that the user would also need to supply a function to 'dump' or stringify the front-matter.

 - Having only YAML and JSON parsers built in seems logical to me. YAML if the default format for front-matter, JSON is probably the next most common (and doesn't require another dependancy).

 - I'm not sure how much this module will be used in a browser, but if we do have a browser compatible version, does it need to support streams?

**Comment by qimingweng** [comment]
@axdg yeah I can see your point for #26, so maybe it makes more sense as a separate npm module? I do think it would be convenient as a tool for novices to come and pick up both front-matter and its reverse.

We can maybe even write a smart YAML re-writer that updates keys in the places they already are and keeps comments as much as possible. But that may be a separate exercise that really does deserve another npm module.

**Comment by jxson** [comment]
@axdg how is the dsfm merge coming along? Is there anything I can do to help?

**Comment by axdg** [comment]
Hey @jxson, apologies, I've had less time to deal with this than I'd expected. 

I've found one minor bug in dsfm (it's not trimming carriage returns (`\r`) where DOS newlines are present) that I'm trying to figure out the fastest method to overcome. Otherwise it's just a matter of changing the names of a few variables and merging both test sets - I will have the v3 branch up and running over the weekend.

**Comment by jxson** [comment]
@axdg no sweat! Just thought I would check in.
